File: contractsdk/cpp/xchain/basic_iterator.cc
#include "xchain/xchain.h"
#include "xchain/basic_iterator.h"
#include "xchain/syscall.h"

namespace pb = xchain::contract::sdk;

namespace xchain {

Iterator::Iterator(const std::string& start, const std::string& limit, size_t cap) {
    _cap = cap;
    _start = start;
    _limit = limit;

    if (!load()) {
        _it = -1;
    } else {
        _it = 0;
    }
}

bool Iterator::load() {
    _buf.clear();
    if (_start == _limit) {
        return false;
    }
    //每次多取一个作为下一次的主键
    bool ok = range_query(_start, _limit, _cap + 1, &_buf);
    if (!ok) {
        error = Error(xchain::ErrorType::kErrIteratorLoad);
        return false;
    }
    if (_buf.size() != 0) {
        if (_buf.size() == _cap + 1) {
            _start = (*(_buf.rbegin())).first;
            _buf.pop_back();
        } else {
            _start = _limit;
            //这个时候就已经结束了. 但是留到下次取的时候返回false
        }
    }
    return true;
}

bool Iterator::next() {
    bool ret = end();
    if (ret) {
        //只有最后一批结束的时候，才会走到这里
        return false;
    }
    _cur_elem = &_buf[_it];
    ++_it;
    if (end()) {
        _last_one = *_cur_elem;
        _cur_elem = &_last_one;
        if (!load()) {
            _it = -1;
        } else {
            _it = 0;
        }
    }
    return !ret;
}

bool Iterator::get(ElemType* t) {
    *t = *_cur_elem;
    return true;
}

bool Iterator::end() {
    return _it >= _buf.size() || _it < 0;
}

bool Iterator::range_query(const std::string& s, const std::string& e,
        const size_t limit, std::vector<std::pair<std::string, std::string>>* res) {
    pb::IteratorRequest req;
    pb::IteratorResponse resp;
    req.set_limit(e);
    req.set_start(s);
    req.set_cap(limit);
    bool ok = syscall("NewIterator", req, &resp);
    if (!ok) {
        return false;
    }
    for (auto item : resp.items()) {
        res->emplace_back(item.key(), item.value());
    }
    return true;
}

}

File: core/contractsdk/java/src/main/java/com/baidu/xuper/BasicIterator.java
package com.baidu.xuper;

import com.baidu.xuper.contractpb.Contract;
import com.baidu.xuper.contractpb.SyscallGrpc;
import com.google.protobuf.ByteString;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;

public class BasicIterator implements Iterator<ContractIterator> {
    static final int CAP = 100;

    private final SyscallGrpc.SyscallBlockingStub client;
    private final Contract.IteratorRequest.Builder builder;
    private byte[] start;
    private byte[] limit;
    private ArrayList<Contract.IteratorItem> items;
    private int it;

    public static BasicIterator newIterator(SyscallGrpc.SyscallBlockingStub client,
                                            Contract.SyscallHeader header, byte[] start, byte[] limit) {
        BasicIterator iter = new BasicIterator(client, header, start, limit);
        if (!iter.load()) {
            iter.it = -1;
        } else {
            iter.it = 0;
        }
        return iter;
    }

    private BasicIterator(SyscallGrpc.SyscallBlockingStub client, Contract.SyscallHeader header,
                          byte[] start, byte[] limit) {
        this.client = client;
        this.builder = Contract.IteratorRequest.newBuilder().setHeader(header);
        this.start = start;
        this.limit = limit;
        this.items = new ArrayList<>();
    }

    @Override
    public boolean hasNext() {
        boolean ret = end();
        if (ret) {
            return false;
        }
        return true;
    }

    @Override
    public ContractIterator next() {
        boolean ret = end();
        if (ret) {
            throw new NoSuchElementException();
        }
        Contract.IteratorItem curItem = this.items.get(this.it);
        this.it++;
        if (end()) {
            if (!load()) {
                this.it = -1;
            } else {
                this.it = 0;
            }
        }

        ContractIterator item = new ContractIterator(curItem);

        return item;
    }

    @Override
    public void remove() {
        throw new UnsupportedOperationException("The BasicIterator does not implement the remove method");
    }

    private boolean load() {
        this.items.clear();
        if (Arrays.equals(this.start, this.limit)) {
            return false;
        }
        Contract.IteratorResponse iteratorResponse = rangeQuery(this.start, this.limit, CAP + 1);
        for (int i = 0; i < iteratorResponse.getItemsCount(); i++) {
            this.items.add(iteratorResponse.getItems(i));
        }
        if (!this.items.isEmpty()) {
            int len = this.items.size();
            if (len == CAP + 1) {
                this.start = this.items.get(len - 1).getKey().toByteArray();
                this.items.remove(len - 1);
            } else {
                this.start = this.limit;
            }
        }
        return true;
    }

    private boolean end() {
        return this.it >= this.items.size() || this.it < 0;
    }

    private Contract.IteratorResponse rangeQuery(byte[] start, byte[] limit, int cap) {
        Contract.IteratorRequest request = this.builder
                .setStart(ByteString.copyFrom(start)).setLimit(ByteString.copyFrom(limit)).setCap(cap).build();

        return this.client.newIterator(request);
    }
}

class ContractIterator {
    private Contract.IteratorItem item;

    public ContractIterator(Contract.IteratorItem item) {
        this.item = item;
    }

    public String getKey() {
        return this.item.getKey().toStringUtf8();
    }

    public String getValue() {
        return this.item.getValue().toStringUtf8();
    }
}
